openedx/core/djangoapps/plugins/plugin_slots.py

```diff
@@ -0,0 +1,114 @@
+from logging import getLogger
+from typing import Dict, List, Tuple, Callable
+
+from django.apps import AppConfig
+from importlib_metadata import ModuleNotFoundError
+
+from openedx.core.lib.cache_utils import process_cached
+from . import constants, registry
+from .plugin_contexts import get_plugins_view_context
+
+log = getLogger(__name__)
+
+COMMON_ALLOW_LIST = ['request', 'current_url']
+
+
+def get_content_for_slot(project_type: str, slot_namespace: str, slot_name: str, raw_context: Dict) -> str:
+    """
+    Returns a list of additional content for a view slot. Will check if any plugin apps
+    have that view in their view_context_config, and if so will call their
+    selected function to get their context dicts.
+
+    Args:
+        project_type (str): a string that determines which project (lms or studio) the view is being called in. See the
+            ProjectType enum in plugins/constants.py for valid options
+        slot_namespace (str): a string that specifies the namespace for this slot.
+        slot_name (str): a string that determines which slot the plugin will render content for. These are unique for
+            each project type.
+        raw_context (Dict): the unfiltered context available to the internal view.
+    """
+    aggregate_slot_content = u""
+    slot_functions = _get_cached_slot_functions_for_view(project_type, slot_namespace, slot_name)
+
+    # Each view can pass the approved part of the context, in the context itself.
+    context_allow_list = raw_context.get('context_allow_list', [])
+    if context_allow_list == '*':
+        allowed_context = raw_context
+    else:
+        # The request object is allowed by default, and always available.
+        context_allow_list.extend(COMMON_ALLOW_LIST)
+
+        allowed_context = {
+            key: raw_context.get(key)
+            for key in context_allow_list
+        }
+
+    for (slot_function, plugin_name) in slot_functions:
+        try:
+            # Allow plugins to extend the context for other plugins
+            context = get_plugins_view_context(
+                constants.ProjectType.LMS,
+                plugin_name,
+                raw_context,
+            )
+            context.update(allowed_context)
+            plugin_slot_content = slot_function(context)
+            aggregate_slot_content += plugin_slot_content
+
+        except Exception as exc:
+            # We're catching this because we don't want the core to blow up when a
+            # plugin is broken. This exception will probably need some sort of
+            # monitoring hooked up to it to make sure that these errors don't go
+            # unseen.
+            log.exception("Failed to call plugin slot function. Error: %s", exc)
+            continue
+
+    return aggregate_slot_content
+
+
+@process_cached
+def _get_cached_slot_functions_for_view(
+    project_type: str, slot_namespace: str, slot_name: str
+) -> List[Tuple[Callable, str]]:
```

Comment:
Seeing as this is mostly copy/pasted from the plugin_contexts, I'm wondering if we can extract this into a shared function.

Reply:
That was my goal once the approach was approved, however I seem to have forgotten that since it was a long time ago. I will look into it.

Reply:
Done.